x/auth/legacy/v043/store.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/gogo/protobuf/grpc"
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const (
+	delegatorDelegationPath = "/cosmos.staking.v1beta1.Query/DelegatorDelegations"
+	balancesPath            = "/cosmos.bank.v1beta1.Query/AllBalances"
+)
+
+func migrateVestingAccounts(ctx sdk.Context, accounts []types.AccountI, queryServer grpc.Server) ([]types.AccountI, error) {
+	for i := 0; i < len(accounts); i++ {
+		asVesting := vesting(accounts[i])
+		if asVesting == nil {
+			continue
+		}
+
+		account := accounts[i]
+
+		addr := account.GetAddress().String()
+		balance, err := getBalance(
+			ctx,
+			addr,
+			queryServer,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		delegations, err := getDelegatorDelegationsSum(
+			ctx,
+			addr,
+			queryServer,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		asVesting, ok := resetVestingDelegatedBalances(asVesting)
+		if !ok {
+			continue
+		}
+
+		// balance before any delegation includes balance of delegation
+		for _, coin := range delegations {
+			balance = balance.Add(coin)
+		}
+
+		asVesting.TrackDelegation(ctx.BlockTime(), balance, delegations)
+
+		accounts[i] = asVesting.(types.AccountI)
+	}
+
+	return accounts, nil
+}
+
+func vesting(account types.AccountI) exported.VestingAccount {
+	v, ok := account.(exported.VestingAccount)
+	if !ok {
+		return nil
+	}
+
+	return v
+}
+
+func resetVestingDelegatedBalances(evacct exported.VestingAccount) (exported.VestingAccount, bool) {
+	// reset `DelegatedVesting` and `DelegatedFree` to zero
+	df := sdk.NewCoins()
+	dv := sdk.NewCoins()
+
+	switch vacct := evacct.(type) {
+	case *vestingtypes.ContinuousVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	case *vestingtypes.DelayedVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	case *vestingtypes.PeriodicVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	default:
+		return nil, false
+	}
+}
+
+func getDelegatorDelegationsSum(ctx sdk.Context, address string, queryServer grpc.Server) (sdk.Coins, error) {
+	querier, ok := queryServer.(*baseapp.GRPCQueryRouter)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type: %T wanted *baseapp.GRPCQueryRouter", queryServer)
+	}
+
+	queryFn := querier.Route(delegatorDelegationPath)
+
+	q := &stakingtypes.QueryDelegatorDelegationsRequest{
+		DelegatorAddr: address,
+	}
+
+	b, err := proto.Marshal(q)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal staking type query request, %w", err)
+	}
+	req := abci.RequestQuery{
+		Data: b,
+		Path: delegatorDelegationPath,
+	}
+	resp, err := queryFn(ctx, req)
```

Comment:
I agree with @amaurym [comment](https://github.com/cosmos/cosmos-sdk/pull/8865#discussion_r598755717). 

We can't add `BankKeeper, StakingKeeper` as direct dependency, but we can:
* create a new structure: `MigratorKeeprs` in ` x/auth/legacy/v043` or in `x/auth/migrations` (in the latter case I would call it Keepers to avoid name clash)
```
type MigratorKeepers struct 
    bank bank.Keeper
    staking staking.Keeper
    //NOTE  we can't put auth Keepers here - this module can't import auth
}
```

* In `NewMigrator`, use `MigratorKeeper` as an argument - since it in a separate package, it won't cause import cycle problem
* pass keepers directly to ` x/auth/legacy/v043` - in this module it should be possible to import staking and bank. 

Other solution is to create (copy) a subset of bank and staking interface needed for migration. That interface can be stored in `auth/legacy/v043` or in `auth/types`. Then we use the same interface as an argument to this function (`getDelegatorDelegationsSum`)

Reply:
> pass keepers directly to x/auth/legacy/v043 - in this module it should be possible to import staking and bank

How do we get staking and bank keeper references? We just create them?

Reply:
It can be a dependency in `AppModule`. Look at `x/staking/module.go`  -- this one already has bank.Keeper. Here, if we have a import cycle problem, we can use an interface subset as I described above.

Reply:
Sorry but this is the same as adding direct dependency on the keepers, isnt it? looks like too much indirection to in the end just adding the keepers for the only purpose of the migration in the AppModule.

Correct me if I am wrong.

Reply:
No, you add dependency to interface. The dependency will be resolved through the keeper, but on a different level. Other solution is related to @amaurym related to query client (without doing any routing).

Reply:
I'd rather leave it as it is right now, and work something out as soon as we have a proper way of doing infra-package communication.

Reply:
Unresolving this comment as this is not solved. I will create an issue for this (we can do it in other PR to not block the release).